Decode XML::LibXML objects through the XS pointer table

sv_setref_pv stores Fil-C pointers as zptrtable-encoded integers, but
XML::LibXML converted them back with INT2PTR(T, SvIV(...)), and stored
its proxy-node registry with PTR2IV. Decode and encode through
Perl_xsub_ptrtable. Generated by rewriting every INT2PTR(T, SvIV(x)).

diff --git a/LibXML.xs b/LibXML.xs
index 8709421..3d2b1fd 100644
--- a/LibXML.xs
+++ b/LibXML.xs
@@ -2796,7 +2796,7 @@ _default_catalog( self, catalog )
         SV * catalog
     PREINIT:
 #ifdef LIBXML_CATALOG_ENABLED
-        xmlCatalogPtr catal = INT2PTR(xmlCatalogPtr,SvIV(SvRV(catalog)));
+        xmlCatalogPtr catal = ((xmlCatalogPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(catalog))));
 #endif
     INIT:
         if ( catal == NULL ) {
@@ -5317,7 +5317,7 @@ _toStringC14N(self, comments=0, xpath=&PL_sv_undef, exclusive=0, inc_prefix_list
                 refNode = xmlDocGetRootElement( self->doc );
             }
 	    if (SvOK(xpath_context)) {
-	      child_ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(xpath_context)));
+	      child_ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(xpath_context))));
 	      if ( child_ctxt == NULL ) {
 		croak("XPathContext: missing xpath context\n");
 	      }
@@ -5429,7 +5429,7 @@ _find( pnode, pxpath, to_bool )
             croak( "lost node" );
         }
         if (sv_isobject(pxpath) && sv_isa(pxpath,"XML::LibXML::XPathExpression")) {
-             comp = INT2PTR(xmlXPathCompExprPtr,SvIV((SV*)SvRV( pxpath )));
+             comp = ((xmlXPathCompExprPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( pxpath ))));
              if (!comp) XSRETURN_UNDEF;
         } else {
             xpath = nodeSv2C(pxpath, node);
@@ -5547,7 +5547,7 @@ _findnodes( pnode, perl_xpath )
 	  croak( "lost node" );
         }
         if (sv_isobject(perl_xpath) && sv_isa(perl_xpath,"XML::LibXML::XPathExpression")) {
-             comp = INT2PTR(xmlXPathCompExprPtr,SvIV((SV*)SvRV( perl_xpath )));
+             comp = ((xmlXPathCompExprPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( perl_xpath ))));
              if (!comp) XSRETURN_UNDEF;
         } else {
             xpath = nodeSv2C(perl_xpath, node);
@@ -7118,7 +7118,7 @@ void
 DESTROY(self)
         SV * self
     PREINIT:
-        xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
+        xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
     CODE:
         xs_warn( "DESTROY NS" );
         if (ns) {
@@ -7131,7 +7131,7 @@ nodeType(self)
     ALIAS:
         getType = 1
     PREINIT:
-        xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
+        xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
     CODE:
         PERL_UNUSED_VAR(ix);
         RETVAL = ns->type;
@@ -7149,7 +7149,7 @@ declaredURI(self)
         value2 = 5
 	href = 6
     PREINIT:
-        xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
+        xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         xmlChar * href;
     CODE:
         PERL_UNUSED_VAR(ix);
@@ -7166,7 +7166,7 @@ declaredPrefix(self)
 	localname = 1
         getLocalName = 2
     PREINIT:
-        xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
+        xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         xmlChar * prefix;
     CODE:
         PERL_UNUSED_VAR(ix);
@@ -7180,7 +7180,7 @@ SV*
 unique_key( self )
         SV * self
     PREINIT:
-        xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
+        xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         xmlChar* key;
     CODE:
         /* Concatenate prefix and URI with vertical bar dividing*/
@@ -7196,8 +7196,8 @@ _isEqual(self, ref_node)
        SV * self
        SV * ref_node
     PREINIT:
-       xmlNsPtr ns = INT2PTR(xmlNsPtr,SvIV(SvRV(self)));
-       xmlNsPtr ons = INT2PTR(xmlNsPtr,SvIV(SvRV(ref_node)));
+       xmlNsPtr ns = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
+       xmlNsPtr ons = ((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(ref_node))));
     CODE:
        RETVAL = 0;
        if ( ns == ons ) {
@@ -7712,7 +7712,7 @@ void
 DESTROY( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
     CODE:
         xs_warn( "DESTROY XPATH CONTEXT" );
         if (ctxt) {
@@ -7751,7 +7751,7 @@ SV*
 getContextNode( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7768,7 +7768,7 @@ int
 getContextPosition( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7781,7 +7781,7 @@ int
 getContextSize( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7795,7 +7795,7 @@ setContextNode( self , pnode )
         SV * self
         SV * pnode
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7814,7 +7814,7 @@ setContextPosition( self , position )
         SV * self
         int position
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL )
             croak("XPathContext: missing xpath context\n");
         if ( position < -1 || position > ctxt->contextSize )
@@ -7827,7 +7827,7 @@ setContextSize( self , size )
         SV * self
         int size
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL )
             croak("XPathContext: missing xpath context\n");
         if ( size < -1 )
@@ -7849,7 +7849,7 @@ registerNs( pxpath_context, prefix, ns_uri )
     PREINIT:
         xmlXPathContextPtr ctxt = NULL;
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7873,7 +7873,7 @@ lookupNs( pxpath_context, prefix )
     PREINIT:
         xmlXPathContextPtr ctxt = NULL;
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7887,7 +7887,7 @@ SV*
 getVarLookupData( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7904,7 +7904,7 @@ SV*
 getVarLookupFunc( self )
         SV * self
     INIT:
-        xmlXPathContextPtr ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(self)));
+        xmlXPathContextPtr ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(self))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -7926,7 +7926,7 @@ registerVarLookupFunc( pxpath_context, lookup_func, lookup_data )
         xmlXPathContextPtr ctxt = NULL;
         XPathContextDataPtr data = NULL;
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL )
             croak("XPathContext: missing xpath context\n");
         data = XPathContextDATA(ctxt);
@@ -7973,7 +7973,7 @@ registerFunctionNS( pxpath_context, name, uri, func)
         char *strkey;
 
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -8036,7 +8036,7 @@ _free_node_pool( pxpath_context )
     PREINIT:
         xmlXPathContextPtr ctxt = NULL;
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -8060,7 +8060,7 @@ _findnodes( pxpath_context, perl_xpath )
         xmlXPathCompExprPtr comp = NULL;
         PREINIT_SAVED_ERROR
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -8069,7 +8069,7 @@ _findnodes( pxpath_context, perl_xpath )
             croak("XPathContext: lost current node\n");
         }
         if (sv_isobject(perl_xpath) && sv_isa(perl_xpath,"XML::LibXML::XPathExpression")) {
-             comp = INT2PTR(xmlXPathCompExprPtr,SvIV((SV*)SvRV( perl_xpath )));
+             comp = ((xmlXPathCompExprPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( perl_xpath ))));
              if (!comp) XSRETURN_UNDEF;
         } else {
             xpath = nodeSv2C(perl_xpath, ctxt->node);
@@ -8164,7 +8164,7 @@ _find( pxpath_context, pxpath, to_bool )
         xmlXPathCompExprPtr comp = NULL;
         PREINIT_SAVED_ERROR
     INIT:
-        ctxt = INT2PTR(xmlXPathContextPtr,SvIV(SvRV(pxpath_context)));
+        ctxt = ((xmlXPathContextPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(pxpath_context))));
         if ( ctxt == NULL ) {
             croak("XPathContext: missing xpath context\n");
         }
@@ -8173,7 +8173,7 @@ _find( pxpath_context, pxpath, to_bool )
             croak("XPathContext: lost current node\n");
         }
         if (sv_isobject(pxpath) && sv_isa(pxpath,"XML::LibXML::XPathExpression")) {
-             comp = INT2PTR(xmlXPathCompExprPtr,SvIV((SV*)SvRV( pxpath )));
+             comp = ((xmlXPathCompExprPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( pxpath ))));
              if (!comp) XSRETURN_UNDEF;
         } else {
             xpath = nodeSv2C(pxpath, ctxt->node);
diff --git a/perl-libxml-mm.c b/perl-libxml-mm.c
index ec2b5ea..b88e9f6 100644
--- a/perl-libxml-mm.c
+++ b/perl-libxml-mm.c
@@ -299,7 +299,7 @@ PmmCloneProxyNodes()
 	xmlHashTablePtr reg_copy;
 	SvLOCK(PROXY_NODE_REGISTRY_MUTEX);
 	reg_copy = xmlHashCopy(PmmREGISTRY, PmmRegistryHashCopier);
-	SvIV_set(SvRV(sv_reg), PTR2IV(reg_copy));
+	SvIV_set(SvRV(sv_reg), zptrtable_encode(Perl_xsub_ptrtable, reg_copy));
 	SvUNLOCK(PROXY_NODE_REGISTRY_MUTEX);
 }
 
diff --git a/perl-libxml-mm.h b/perl-libxml-mm.h
index aaeb719..cb4480f 100644
--- a/perl-libxml-mm.h
+++ b/perl-libxml-mm.h
@@ -137,9 +137,9 @@ typedef ProxyNode* ProxyNodePtr;
 typedef DocProxyNode* DocProxyNodePtr;
 
 /* this my go only into the header used by the xs */
-#define SvPROXYNODE(x) (INT2PTR(ProxyNodePtr,SvIV(SvRV(x))))
+#define SvPROXYNODE(x) (((ProxyNodePtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(x)))))
 #define PmmPROXYNODE(x) (INT2PTR(ProxyNodePtr,x->_private))
-#define SvNAMESPACE(x) (INT2PTR(xmlNsPtr,SvIV(SvRV(x))))
+#define SvNAMESPACE(x) (((xmlNsPtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(x)))))
 
 #define PmmREFCNT(node)      node->count
 #define PmmREFCNT_inc(node)  node->count++
@@ -177,8 +177,8 @@ typedef struct _LocalProxyNode LocalProxyNode;
 typedef LocalProxyNode* LocalProxyNodePtr;
 
 #define PmmUSEREGISTRY		(PROXY_NODE_REGISTRY_MUTEX != NULL)
-#define PmmREGISTRY		(INT2PTR(xmlHashTablePtr,SvIV(SvRV(get_sv("XML::LibXML::__PROXY_NODE_REGISTRY",0)))))
-/* #define PmmREGISTRY			(INT2PTR(xmlHashTablePtr,SvIV(SvRV(PROXY_NODE_REGISTRY)))) */
+#define PmmREGISTRY		(((xmlHashTablePtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(get_sv("XML::LibXML::__PROXY_NODE_REGISTRY",0))))))
+/* #define PmmREGISTRY			(((xmlHashTablePtr)zptrtable_decode(Perl_xsub_ptrtable, SvIV(SvRV(PROXY_NODE_REGISTRY))))) */
 
 void
 PmmCloneProxyNodes();
diff --git a/typemap b/typemap
index b10104b..175e437 100644
--- a/typemap
+++ b/typemap
@@ -18,7 +18,7 @@ xmlXPathCompExprPtr         O_XPATH_OBJECT
 INPUT
 O_OBJECT
     if( sv_isobject($arg) && (SvTYPE(SvRV($arg)) == SVt_PVMG) )
-            $var = INT2PTR($type,SvIV((SV*)SvRV( $arg )));
+            $var = (($type)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( $arg ))));
     else{
             warn( \"${Package}::$func_name() -- $var is not a blessed SV reference\" );
             XSRETURN_UNDEF;
@@ -65,7 +65,7 @@ O_PARSER_OBJECT
 
 O_XPATH_OBJECT
     if( sv_isobject($arg) && (SvTYPE(SvRV($arg)) == SVt_PVMG) && sv_isa($arg,\"XML::LibXML::XPathExpression\"))
-            $var = INT2PTR($type,SvIV((SV*)SvRV( $arg )));
+            $var = (($type)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( $arg ))));
     else{
             warn( \"${Package}::$func_name() -- $var is not a XML::LibXML::XPathExpression\" );
             XSRETURN_UNDEF;
@@ -73,7 +73,7 @@ O_XPATH_OBJECT
 
 O_PATTERN_OBJECT
     if( sv_isobject($arg) && (SvTYPE(SvRV($arg)) == SVt_PVMG) && sv_isa($arg,\"XML::LibXML::Pattern\"))
-            $var = INT2PTR($type,SvIV((SV*)SvRV( $arg )));
+            $var = (($type)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( $arg ))));
     else{
             warn( \"${Package}::$func_name() -- $var is not a XML::LibXML::Pattern\" );
             XSRETURN_UNDEF;
@@ -81,7 +81,7 @@ O_PATTERN_OBJECT
 
 O_REGEXP_OBJECT
     if( sv_isobject($arg) && (SvTYPE(SvRV($arg)) == SVt_PVMG) && sv_isa($arg,\"XML::LibXML::RegExp\"))
-            $var = INT2PTR($type,SvIV((SV*)SvRV( $arg )));
+            $var = (($type)zptrtable_decode(Perl_xsub_ptrtable, SvIV((SV*)SvRV( $arg ))));
     else{
             warn( \"${Package}::$func_name() -- $var is not a XML::LibXML::RegExp\" );
             XSRETURN_UNDEF;
